server/paymentProviders/opencollective/virtualcard.js

```diff
@@ -0,0 +1,181 @@
+ *  virtual card payment method.
+ * @return {Object} with amount & currency from the payment method.
+ */
+async function getBalance(paymentMethod) {
+  if (!libpayments.isProvider('opencollective.virtualcard', paymentMethod)) {
+    throw new Error(`Expected opencollective.virtualcard but got ${paymentMethod.service}.${paymentMethod.type}`);
+  }
+  /* Result will be negative (We're looking for DEBIT transactions) */
+  const spent = await libtransactions.sum({
+    PaymentMethodId: paymentMethod.id,
+    currency: paymentMethod.currency,
+    type: 'DEBIT',
+  });
+  return {
+    amount: paymentMethod.initialBalance + spent,
+    currency: paymentMethod.currency,
+  };
+}
+
+/** Process a virtual card order
+ *
+ * @param {models.Order} order The order instance to be processed.
+ * @return {models.Transaction} the double entry generated transactions.
+ */
+async function processOrder(order) {
+  const paymentMethod = await models.PaymentMethod.findById(order.paymentMethod.id);
+  // check if payment Method has expired
+  if (!paymentMethod.expiryDate || moment(paymentMethod.expiryDate) < moment()) {
+    throw new Error('Payment method has already expired');
+  }
+
+  // Checking if balance is ok or will still be after completing the order
+  const balance = await getBalance(paymentMethod);
+  if (!balance || balance.amount <= 0) {
+    throw new Error('Virtual card has no balance to complete this order');
+  }
+  if ( (balance.amount - order.totalAmount) < 0 ) {
+    throw new Error(`Order amount exceeds balance(${balance.amount} ${paymentMethod.currency})`);
+  }
+
+  // Making sure the SourcePaymentMethodId is Set(requirement for virtual cards)
+  if (!get(paymentMethod, 'SourcePaymentMethodId')) {
+    throw new Error('Virtual Card payment method must have a value a "SourcePaymentMethodId" defined');
+  }
+  // finding Source Payment method and update order payment method properties
+  const sourcePaymentMethod = await models.PaymentMethod.findById(paymentMethod.SourcePaymentMethodId);
+  order.PaymentMethodId = sourcePaymentMethod.id;
+  order.paymentMethod = sourcePaymentMethod;
+  // finding the payment provider lib to execute the order
+  const sourcePaymentMethodProvider = libpayments.findPaymentMethodProvider(sourcePaymentMethod);
+
+  // gets the Credit transaction generated
+  const creditTransaction = await sourcePaymentMethodProvider.processOrder(order);
+  // gets the Debit transaction generated through the TransactionGroup field.
+  const debitTransaction = await models.Transaction.findOne({
+    where: {
+      type: 'DEBIT',
+      TransactionGroup: creditTransaction.TransactionGroup,
+    },
+  });
+  // Updating already created transactions to use the Virtual Payment Method id instead
+  const updatedPaymentMethodTransactions =
+    await Promise.map([creditTransaction, debitTransaction], async (transaction) => {
+      await transaction.update({ PaymentMethodId: paymentMethod.id });
+      return transaction;
+  });
```

Comment:
Don't fetch the debit transaction, instead just update both based on their group:

```js
await models.PaymentMethod.update({ PaymentMethodId: paymentMethod.id }, {
  where: { TransactionGroup: creditTransaction.TransactionGroup }
});
```